src/es-module-shims.js

```diff
@@ -386,13 +389,23 @@ function processPreload (link) {
   link.ep = true;
   // prepopulate the load record
   const fetchOpts = getFetchOpts(link);
-  // save preloaded fetch options for later load  
+  // save preloaded fetch options for later load
   fetchOptsMap.set(link.href, fetchOpts);
   fetch(link.href, fetchOpts);
 }
 
-function resolve (id, parentUrl) {
-  const urlResolved = resolveIfNotPlainOrUrl(id, parentUrl);
+
+async function resolve(id, parentUrl) {
+  let urlResolved;
+
+  if (esmsInitOptions.resolve) {
+    urlResolved = await esmsInitOptions.resolve(id, parentUrl, resolveIfNotPlainOrUrl);
+  }
+
+  if (!urlResolved) {
+    urlResolved = resolveIfNotPlainOrUrl(id, parentUrl);
+  }
+
   const resolved = resolveImportMap(importMap, urlResolved || id, parentUrl);
   return { r: resolved, m: urlResolved !== resolved };
```

Comment:
I know it's not necessarily clear what the purpose of urlResolved is, but the following should better maintain the current intent:

```suggestion
  const urlResolved = resolveIfNotPlainOrUrl(id, parentUrl);

  let resolved;
  if (esmsInitOptions.resolve) {
    resolved = await esmsInitOptions.resolve(id, parentUrl, resolveIfNotPlainOrUrl);
  }
  else {
    resolved = resolveImportMap(importMap, resolved || id, parentUrl);
  }

  return { r: resolved, m: urlResolved !== resolved };
```

Reply:
Finally I think we should pass the full original resolver as the third argument to `esmsInitOptions.resolve` not just the `resolveIfNotPlainOrUrl` - that way the resolver hook has access to import map resolutions.

Reply:
Pass all context to the third argument?
```js
esmsInitOptions.resolve(id, parentUrl, { resolveIfNotPlainOrUrl, resolveImportMap, importMap });
```
So the resolver hook can reuse this?